symengine/functions.cpp

```diff
@@ -3571,4 +3620,23 @@ RCP<const Basic> min(const vec_basic &arg)
     }
 }
 
+RCP<const Basic> expand_as_exp(const RCP<const Basic> &self)
+{
+    map_basic_basic mp_args;
+    for (const auto &a : self->get_args()) {
+        mp_args[a] = expand_as_exp(a);
```

Comment:
As @isuruf pointed out [here](https://github.com/symengine/symengine/pull/1305#discussion_r126961094), This is inefficient.
I tried to implement what he suggested there. I was just waiting for his confirmation on the following design.

```
template <typename Func>
class BottomUpVisitor : public BaseVisitor<BottomUpVisitor<Func>>
{
private:
    RCP<const Basic> result_;
    Func f;

public:
    BottomUpVisitor(const Func &f_) : f(f_){}
};

inline RCP<const Basic> identity(const RCP<const Basic> &arg) { return arg; }

class ExpandasExp : public BaseVisitor<ExpandasExp, BottomUpVisitor<decltype(&identity)>>
{
public:
    using BottomUpVisitor::bvisit;

    ExpandasExp() : BaseVisitor<ExpandasExp, BottomUpVisitor<decltype(&identity)>>(&identity){}

    template<typename T> // Overload for Sin etc..
    void bvisit(const T &x)
    {
        ...
        result_ = x.expand_as_exp();
        ...
    }
};
```
Btw, I could see visible speed diff in both(~7-10X faster :o ).